caput/mpiarray.py

```diff
@@ -214,15 +214,23 @@ def __getitem__(self, slobj):
             slobj = tuple(slice(None, None, None) if sl is None else sl for sl in slobj)
 
             # Return an MPIArray view
-            arr = self.array[slobj]
-
             # Figure out which is the distributed axis after the slicing, by
             # removing slice axes which are just ints from the mapping
             dist_axis = [
                 index for index, sl in enumerate(slobj) if not isinstance(sl, int)
             ].index(self.axis)
 
-            return MPIArray.wrap(arr, axis=dist_axis, comm=self.array._comm)
+            if dist_axis != self.axis:
+
+                return MPIArray.wrap(
+                    self.array.view(np.ndarray)[slobj],
+                    axis=dist_axis,
+                    comm=self.array._comm,
+                )
+            else:
+                return MPIArray.wrap(
+                    self.array[slobj], axis=dist_axis, comm=self.array._comm
```

Comment:
It is better to reduce the number of hidden MPI calls, and so it would be better to do this within an MPIArray `__getitem__`, and remove `wraps`. If the issue is the calculation of the `totallength` of the distributed axes, we can prevent operations which adjust the size of the total length, and grab it from the global shape of the original axes..